### PR TITLE
US024 - Registration System

### PR DESCRIPTION
**Pull request by LMBishop**
Corresponding frontend PR: https://github.com/anthony-raybould/bham1-front-end/pull/17

This PR adds two endpoints for registration: `register` to create an account, and `roles` to retrieve role IDs and names for the frontend.

image
image

### CODE REVIEW

File: src/main/java/org/kainos/ea/api/AuthService.java [hunk 1]
```diff
@@ -37,4 +37,26 @@ public String login(Login login) throws FailedToGenerateTokenException, FailedTo
...
         }
     }
+
+    public void register(RegisterRequest request) throws FailedToRegisterException, DuplicateRegistrationException {
+        String salt = BCrypt.gensalt(10);
+        String hashedPassword = BCrypt.hashpw(request.getPassword(), salt);
+        try {
+            authDao.register(request.getEmail(), hashedPassword, request.getRole());
+        } catch (SQLException e) {
+            if (e.getErrorCode() == 1062) {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/api/AuthService.java]: Perfect place for a enum of SQL Error Codes

File: src/main/java/org/kainos/ea/resources/AuthController.java [hunk 2]
```diff
@@ -59,6 +56,24 @@ public Response login(Login login) {
...
+    @POST
+    @Path("/register")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = "Registers a user")
+    public Response register(RegisterRequest request) {
+        try {
+            authService.register(request);
+
+            return Response.status(Response.Status.CREATED).build();
+        } catch (FailedToRegisterException e) {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/resources/AuthController.java]: This exception could be thrown for any issue from the DB, it doesn't mean that it was down to user input, so 400 doesn't feel right here

File: src/main/java/org/kainos/ea/api/AuthService.java [hunk 3]
```diff
@@ -37,4 +37,26 @@ public String login(Login login) throws FailedToGenerateTokenException, FailedTo
             throw new FailedToGenerateTokenException();
         }
     }
+
+    public void register(RegisterRequest request) throws FailedToRegisterException, DuplicateRegistrationException {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/api/AuthService.java]: Missing validation on the request. I could submit anything via swagger/api

File: src/test/java/org/kainos/ea/dao/AuthDaoTest.java [hunk 4]
```diff
@@ -82,4 +87,26 @@ public void getUserId_shouldReturnUserId_whenValid() throws SQLException {
...
+        Mockito.when(resultSetMock.getInt("roleID")).thenReturn(1);
+        Mockito.when(resultSetMock.getString("name")).thenReturn("Admin");
+
+        List<Role> roles = authDao.getRoles();
+
+        assertEquals(roles.size(), 1);
+        assertEquals(roles.get(0).getRoleId(), 1);
+        assertEquals(roles.get(0).getRoleName(), "Admin");
+    }
+
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/dao/AuthDaoTest.java:109]: I don't see any tests for registration

File: src/main/java/org/kainos/ea/resources/AuthController.java [hunk 5]
```diff
@@ -69,4 +84,21 @@ public Response login(Login login) {
     public Response whoami(@Auth @ApiParam(hidden = true) User user) {
         return Response.ok().entity(new UserResponse(user)).build();
     }
+
+    @GET
+    @Path("/roles")
+    @PermitAll
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = "Returns a list of available user roles", response = RoleResponse.class, responseContainer = "List")
+    public Response roles() {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/resources/AuthController.java:95]: Missing integration tests

File: src/main/java/org/kainos/ea/resources/AuthController.java [hunk 6]
```diff
@@ -59,6 +56,24 @@ public Response login(Login login) {
...
+        try {
+            authService.register(request);
+
+            return Response.status(Response.Status.CREATED).build();
+        } catch (FailedToRegisterException e) {
+            System.err.println(e.getMessage());
+
+            return Response.status(Response.Status.BAD_REQUEST).entity(e.getMessage()).build();
+        } catch (DuplicateRegistrationException e) {
+            return Response.status(Response.Status.CONFLICT).entity(e.getMessage()).build();
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/resources/AuthController.java:70]: 👍

File: src/test/java/org/kainos/ea/integration/RegisterIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @Test
+    void register_shouldReturn409_whenDuplicate() {
+        Response response = APP.client().target(System.getenv("TARGET_DOMAIN") + "/api/register").request()
+                .post(Entity.entity(duplicateRequest, MediaType.APPLICATION_JSON));
+        Assertions.assertEquals(409, response.getStatus());
+    }
+
+}
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/integration/RegisterIntegrationTest.java:65]: You probably want 400 test(s) as well

**Review by TomBillingtonUK — APPROVED** [review]

**Review by ManpreetM01 — APPROVED** [review]

**Pull request merged**